[Core] Add pwned passwords API to prevent common passwords and password reuse

johnsaigle opened the pull request:
### Brief summary of changes

Add the [pwned passwords API](https://haveibeenpwned.com/API/v2#PwnedPasswords) into the LORIS password logic. This is part of a refactoring effort to modernize the way LORIS does passwords.

In particular, this API prevents users from creating passwords that are known to be present in previous data leaks. After a password breach, hackers compose large lists of the password and try them in brute force attacks against the sites. 

After this PR users will be prevented from creating a password like `Spring2016!` which is a bad password even though it passes the (current) LORIS password strength requirements.

### Related to

#3314, #3949, #3950 

### To test this change...

- [ ] Checkout my branch.
- [ ] Run the new patch on your DB: `SQL/New_patches/2019-01-28_Add_Pwned_Password_ConfigSetting.sql`
- [ ] On your VM go to the Configuration menu and enable the "Use Pwned Password" check
- [ ] Try changing your password to `Spring2016!`. You should get an error message. 

As @zaliqarosli pointed out, curl is required so if you get an error do 
`sudo apt-get install php7.2-curl; sudo service apache2 restart`

PapillonMcGill commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +525,55 @@ class User extends UserPermissions
+         * the colon and count from the end of the $suffix line.
+         * Check the above link if you want more info.
+         */
+        foreach ($hashSuffixes as $suffix) {
```
What about a password like 'vDg62fF6$gN3&dsKe!dfw' that might have been pwn only once? If we remove all PWD that had been pwn, we will end up blocking a lot of them and eventually it will be difficult to find one that have not been pwn.
A quick extrapolation with 3 test would eliminate about 189M passwords, this would just increased in size over time.

Maybe combine the number of pwn and strength together

johnsaigle replied:
I think the example you gave is a bit of an edge case, and someone with that password is likely to be using a password generator to create it. It's easy to click the button again and generate a new string of random characters.

Even eliminating 189M doesn't put a dent in the entropy of possible passwords. The end goal is to allow any password that is greater than 8 characters that consists of any printable ASCII characters and isn't silly. There's plenty of room to make a good password with that criteria.

PapillonMcGill commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +525,55 @@ class User extends UserPermissions
+         */
+        $hash       = sha1($password);
+        $hashPrefix = substr(sha1($password), 0, 5);
+        $url        = "https://api.pwnedpasswords.com/range/$hashPrefix";
```
Need to add a note in the documentation that need access to outside site. Lots of network admin will prevent outside access from web server to prevent contact to a C&C.

Need to manage timeout in case network admin refuse to open firewall or site is down.
Caveat, some firewall require IP address and do not support FQDN.

PapillonMcGill replied:
Sorry for double post to add in the comment, GitHub is quite slow to update a page with latest comment this morning

johnsaigle commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +525,55 @@ class User extends UserPermissions
+         */
+        $hash       = sha1($password);
+        $hashPrefix = substr(sha1($password), 0, 5);
+        $url        = "https://api.pwnedpasswords.com/range/$hashPrefix";
```
I'll add a comment

johnsaigle commented:
@PapillonMcGill I added code comments to address the firewall case. I also added error handling in the code so that password changes can continue even if the API is not reachable.

PapillonMcGill commented:
@johnsaigle For the note in the documentation, I was thinking more in the Loris install notes (README.md).

Apart that, the code is fine

johnsaigle commented:
I'm trying to find an appropriate place in the documentation but I'm not sure where it should go. There doesn't seem to be a Wiki page really suited for it and it feels too detailed for the main README.md.

Do you have any thoughts on this?

PapillonMcGill commented:
in https://github.com/aces/Loris/wiki/Setup#installation-setup-and-configuration, 10. Opening firewall to ... and create a page for it.

As the wiki is not tied to any PR or release, I will approve this PR.

zaliqarosli commented:
from admin account, i did not get an error message when changing a user's password to `Spring2016!`. from user account, I get a 500 internal server error when attempting to save a new password

johnsaigle commented:
@zaliqarosli I wasn't able to reproduce those errors but I did realize that I hadn't added the pwned check to edit_user as I was testing in my_preferences.

Would you mind testing again? If it still doesn't work, could you show me your error logs?

zaliqarosli commented:
`PHP Fatal error:  Uncaught Error: Call to undefined function curl_init() in /var/www/loris/php/libraries/User.class.inc:553\nStack trace:\n#0 /var/www/loris/modules/user_accounts/php/edit_user.class.inc(1033): User::pwnedPassword('Spring2016!')\n#1 /var/www/loris/php/libraries/LorisForm.class.inc(1322): LORIS\\user_accounts\\Edit_User->_validateEditUser(Array)\n#2 /var/www/loris/php/libraries/NDB_Form.class.inc(132): LorisForm->validate()\n#3 /var/www/loris/php/libraries/NDB_Form.class.inc(190): NDB_Form->save()\n#4 /var/www/loris/modules/user_accounts/php/edit_user.class.inc(562): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#5 /var/www/loris/src/Middleware/UserPageDecorationMiddleware.php(171): LORIS\\user_accounts\\Edit_User->handle(Object(Zend\\Diactoros\\ServerRequest))\n#6 /var/www/loris/src/Middleware/PageDecorationMiddleware.php(49): LORIS\\Middleware\\UserPageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#7 /var/www/loris/php/libraries/NDB_Page.class.inc(650): LO in /var/www/loris/php/libraries/User.class.inc on line 553, referer: http://zrosli-dev.loris.ca/user_accounts/edit_user/admin
`

johnsaigle commented:
thanks @zaliqarosli! Curl is required. I added instructions to the description

johnsaigle commented:
@PapillonMcGill I think your approval got removed by my last couple commits. When this passes Travis, could you review quickly and approve again?

christinerogers commented:
Given `sudo apt-get install` - what is required for CentOS installs?

johnsaigle commented:
@christinerogers I think the equivalent would be using `yum` instead of `apt` but I don't have a CentOS server to test on. Also this dependency may be handled by `composer` in the future if we move to PSR18 as Dave mentioned.

johnsaigle commented:
Adding blocked because this probably won't be merged until we use a PSR18 implementation

PapillonMcGill commented:
@johnsaigle @christinerogers 
`yum install <package-name>`  will work for CentOS

nicolasbrossard commented:
@johnsaigle I like the idea of checking if a password was pwned. I do notice though that the logic for checking if a password change is legit (namely checking if the new password is "strong", different from the old password and not pwned) is duplicated in a couple of classes. We could encapsulate the algorithm in the `User` class and have a method like `isPasswordChangeValid(old, new)` that would execute the above mentioned checks. Maybe even have a `Password` class that will contain everything password-related? Up to you to see what is more appropriate.

johnsaigle commented:
@nicolasbrossard I agree with you that there is unnecessary duplication. I would be in favour of merging this PR and later on I could write a new PR that does exactly what you've said. Since this PR has been open for 4 months already, I am hesitant to make it more complex at this stage.

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,63 @@ class User extends UserPermissions
+        );
+        $request  = new Request('GET', "/range/$hashPrefix");
+        $response = $adapter->sendRequest($request);
+        print_r($response);
```
Leftover debug statement
```suggestion
```

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,63 @@ class User extends UserPermissions
+     */
+    function pwnedPassword(string $password): bool
+    {
+        // TODO Use the HttpClient in LORIS core to cURL more simply.
```
I don't think this comment is valid anymore.

driusan commented:
I agree with @johnsaigle, we can make an issue about that instead of blocking this PR over it.

johnsaigle commented:
@driusan I prefer commenting out code with `#` because it's one less character and it won't get confused with real `/*` or `//` comments. :laughing:

nicolasbrossard commented:
@driusan @johnsaigle fine by me to do the modifications I requested later on. PR LGTM then.

johnsaigle commented:
@driusan @nicolasbrossard Could you give an approving review if satisfied with the code?

zaliqarosli commented:
tested user accounts as described in the PR description and it works! did not test single point login however. not sure if i should pass manual testing

johnsaigle commented:
@zaliqarosli An easy way to test the SinglePointLogin approach is to expire your current password in the database by setting `Password_expiry` to some date in the past. Thanks for testing!

zaliqarosli commented:
update: single point login allows pwned passwords to pass. this needs fixing

johnsaigle commented:
@zaliqarosli thanks for testing! Should be fixed now

PapillonMcGill commented:
LGTM after Travis validation.

zaliqarosli commented:
@johnsaigle not sure how to test the single point login if the password expiry case is covered my the login module. let me know!

johnsaigle commented:
@zaliqarosli I'm not actually sure how to trigger that case. I think a lot of this is left over from the old way we used to do logins and may not actually be reachable code... I think it should be tested and refactored as Nicolas said above. 

Since the pwnedPassword check does not break password changes either way, I think it's OK to merge this if your other tests worked well.

zaliqarosli commented:
tested all cases but single point login and they all work. passing manual test assuming that single point login is deprecated

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,60 @@ class User extends UserPermissions
+            return false;
+        }
+        $hashSuffixes = explode("\n", $response->getBody());
+        /* The response will be one long string consisting of suffixes matching
```
```suggestion
        /*
         * The response will be one long string consisting of suffixes matching
```

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,60 @@ class User extends UserPermissions
+        $request  = new Request('GET', "/range/$hashPrefix");
+        $response = $adapter->sendRequest($request);
+
+        /* We don't want password changes to fail in case there is a
```
```suggestion
        /*
         * We don't want password changes to fail in case there is a
```

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,60 @@ class User extends UserPermissions
+     */
+    function pwnedPassword(string $password): bool
+    {
+        /* The pwned passwords API searches a range of password hashes to
```
```suggestion
        /*
         * The pwned passwords API searches a range of password hashes to
```

johnsaigle replied:
Is it worth adding a PHPCS rule for this? I didn't know this was a style thing we had.

driusan commented on php/libraries/User.class.inc:
```diff
@@ -525,6 +527,60 @@ class User extends UserPermissions
+     *
+     * @return bool Whether the password is known to be pwned.
+     */
+    function pwnedPassword(string $password): bool
```
I think this should start with a config option check to disable this so that LORIS isn't unuseable at places that either don't want or can't allow access to external sites.

johnsaigle replied:
Line 562 makes sure that the password change will continue if for some reason the pwnedPassword API is unreachable, including networking configs or outage of the site. 

Given that LORIS should still be usable with this check, would you still prefer an explicit config option? If so, should that be something in `config.xml` or?

driusan replied:
Without an explicit config check, it requires an attempt to contact the server to error out before it response to the user's request (which could take on the order of tens of seconds depending on the reason for the failure), even with that fallback. It's also possible that some users just won't want to use the external service, even though it's available.

I think it should go in the config tables, since we're trying to phase out config.xml

driusan commented on modules/user_accounts/php/my_preferences.class.inc:
```diff
@@ -429,6 +429,7 @@ class My_Preferences extends \NDB_Form
 
         // if password is user-defined, and user wants to change password
         if (!empty($values['Password_hash'])) {
+
```
```suggestion
```

johnsaigle commented:
@driusan Can you update your review?

driusan commented on php/libraries/User.class.inc:
```diff
@@ -11,6 +11,8 @@
  */
 
+use Http\Adapter\Guzzle6\Client as GuzzleAdapter;
+use GuzzleHttp\Psr7\Request;
```
This should be using PSR-18, not a specific hardcoded implementation of PSR-18.

(To do that you'll probably need to wrap the reference to Guzzle in a factory somewhere.)

johnsaigle replied:
@driusan I just sent a commit that I think does what you asked. Let me know if there's something to correct.

zaliqarosli commented on SQL/New_patches/2019-01-28_Add_Pwned_Password_ConfigSetting.sql:
```diff
@@ -1,6 +1,6 @@
 -- Adds the option to toggle the usage of the Pwned Passwords
 -- API (https://haveibeenpwned.com/API/v2#PwnedPasswords) in a project. This
--- should be enabled by default to allow for a higher level of security. This
+-- is be enabled by default to allow for a higher level of security. This
```
"be" to still be removed

driusan commented on php/libraries/User.class.inc:
```diff
@@ -531,6 +533,70 @@ class User extends UserPermissions
+            'usePwnedPasswordsAPI'
+        )
+        ) {
+            error_log('Setting disabled!');
```
left over debug statement

driusan commented on src/Http/Client.php:
```diff
@@ -0,0 +1,43 @@
+     * Guzzle's Client class is final and so cannot be extended directly. This
+     * function acts as a wrapper for this function and serves as a workaround.
+     */
+    public static function createWithConfig(array $config): GuzzleClient {
```
If a named constructor is required, this should still be returning a Psr\Http\Client\ClientInterface, not a specific implementation of one.

The "config" array is also undocumented here. It seems to only ever have one option passed, so I think it should be `public static function createWithBaseURI(string) : ClientInterface`

This whole class is also acting as a ClientFactory, not a Client.

johnsaigle commented:
Blocked by #4387

kchatpar commented:
Password reset emails are not sent to users because #4362 has not been merged to major yet.

johnsaigle commented:
@kchatpar My bad, my testing instructions weren't clear. Please run the SQL patch `SQL/New_patches/2019-01-28_Add_Pwned_Password_ConfigSetting.sql` in this PR and try again. I'll update the instructions in the PR description.

kchatpar commented:
PR passes the manual test with the password Spring2016! The password spring2016! does not trigger the error message. 

Consider, making a new password rule where one letter in the password must be capitalized.

johnsaigle commented:
Sorry @kchatpar it should do the opposite and I forgot one more testing step. You'll need to go to Admin > Configuration and set "Use Pwned Passwords" to True. 

We have some password rules but they're a little old-fashioned. I have another open PR to update those.

I'll update the testing instructions again.

kchatpar commented:
I set the "Use Pwned Passwords" to True and I still get the same result - Spring2016! is rejected but spring2016! is accepted.

johnsaigle commented:
@kchatpar That's OK for now. This PR addresses previously leaked passwords, not password complexity. That will be handled in #3962.

Thanks for testing!

johnsaigle commented:
Thanks again @zaliqarosli. 

For any new reviewers, password changes can be triggered via edit_user, my_preferences, or via an logging in with an expired password. There is another case that appears in `php/libraries/SinglePointLogin.class.inc` but we're not sure how to trigger that code path. However in all other circumstances the pwned password check works.

This last case will likely be cleaned up as login/password refactoring changes in PRs like #4395 and #4411.
